### PR TITLE
Telemetry Abstractions API Surface

### PR DESCRIPTION
**Pull request by geeknoid**
Please focus on APIs marked [Experimental]

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4375)

### CODE REVIEW

**Comment by danmoseley** [comment]
@JamesNK @noahfalk  any remaining feedback here?

File: API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Telemetry.Logging/LoggerMessageState.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    public readonly struct ClassifiedTag
+    {
+        public string Name { get; }
+        public object? Value { get; }
+        public DataClassification Classification { get; }
+        public ClassifiedTag(string name, object? value, DataClassification classification);
+    }
+    public KeyValuePair<string, object?>[] TagArray { get; }
+    public KeyValuePair<string, object?>[] RedactedTagArray { get; }
+    public ClassifiedTag[] ClassifiedTagArray { get; }
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Telemetry.Logging/LoggerMessageState.cs:27]: This is a helper type for the logging code generator, and not intended to be used by customers directly. The type is documented as such, and the API is marked as EditorBrowsable=Never

**Comment by geeknoid** [comment]
@danmoseley This is intended for the internal R9 API review.

https://github.com/dotnet/extensions/pull/4373 is intended for the .NET team's commentary on stuff, that's why I added all the .NET folks as reviewers there.

File: API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/TelemetryConstants.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Assembly 'Microsoft.Extensions.Telemetry.Abstractions'
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Extensions.Http.Telemetry;
+
+public static class TelemetryConstants
+{
+    public const string RequestMetadataKey = "R9-RequestMetadata";
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/TelemetryConstants.cs:9]: @xakep139

File: API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Assembly 'Microsoft.Extensions.Telemetry.Abstractions'
+
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Http.Telemetry;
+
+public class RequestMetadata
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs:7]: Why? We generally don't seal public types.

File: API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Telemetry.Logging/LoggerMessageState.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public int NumTags { get; }
+    public int NumClassifiedTags { get; }
```
[the remark this answers is not in the document]
Comment (geeknoid) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Telemetry.Logging/LoggerMessageState.cs:29]: This reflects the number of entries present in the larger arrays.

[the remark this answers is not in the document]
Comment (geeknoid) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Telemetry.Logging/LoggerMessageState.cs:29]: https://github.com/dotnet/extensions/pull/4384

File: API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (danmoseley) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs:7]: I believe nowadays we seal wherever we can. It improves devirtualization. @stephentoub ?

[the remark this answers is not in the document]
Comment (stephentoub) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs:7]: I don't see anything virtual here. While I personally advocate for sealing, our design guidelines still say not to by default.

[the remark this answers is not in the document]
Comment (JamesNK) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs:7]: Today I learned. I’ve been sealing types that don’t make sense to inherit from for years.

[the remark this answers is not in the document]
Comment (stephentoub) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs:7]: And I applaud you for doing so.

[the remark this answers is not in the document]
Comment (stephentoub) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/RequestMetadata.cs:7]: (Note there are still perf advantages to sealing even without devirtualization.)

File: API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/TelemetryConstants.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xakep139) [reply API.Microsoft.Extensions.Telemetry.Abstractions/NoDocs/Microsoft.Extensions.Http.Telemetry/TelemetryConstants.cs:9]: Fixed in #4532

**Pull request closed without merging**